### ISSUE TITLE
Smart contract doesn't support object array

### ISSUE THREAD

**Issue by Qiao-Jin** [neo-project/neo-devpack-dotnet#168]
It seems that neo-compiler & neo-vm don't support object array in smart contract currently, i.e.

```
using Neo.SmartContract.Framework;
using System.Numerics;

namespace TestArray
{
    public class Contract1 : SmartContract
    {
        public static object Main(string operation, object[] args)
        {
            State[] states = new State[] { new State() };
            for (var i = 0; i < 1; i++)
            {
                State state = states[i];
            }
            return true;
        }

        public struct State
        {
            public byte[] from;
            public byte[] to;
            public BigInteger amount;
        }
    }
}
```

The simple smart contract will report an error if built in release mode: 

```
IL_000D Ldelema  ---> System.Exception: unsupported instruction Ldelema
```	

After researching I found out this is due to some lack of functionality in func `neo-compiler\neon\MSIL\Converter.ConvertCode`:

```
case CodeEx.Switch:
{
    throw new Exception("need neo.VM update.");
    //var addrdata = new byte[src.tokenAddr_Switch.Length * 2 + 2];
    //var shortaddrcount = (UInt16)src.tokenAddr_Switch.Length;
    //var data = BitConverter.GetBytes(shortaddrcount);
    //addrdata[0] = data[0];
    //addrdata[1] = data[1];
    //var code = _Convert1by1(VM.OpCode.SWITCH, src, to, addrdata);
    //code.needfix = true;
    //code.srcaddrswitch = new int[shortaddrcount];
    //for (var i = 0; i < shortaddrcount; i++)
    //{
    //    code.srcaddrswitch[i] = src.tokenAddr_Switch[i];
    //}
    }
    break;
```	

This piece of code is commented out because VM.OpCode.SWITCH and corresponding logic is not yet defined. By blaming the code it seems this piece of code is inserted in 2017. I suppose this pending functionality might be forgotten?

Furthermore, even if it can be built successfully under debug mode, still an error would occur if run on chain. After debugging it seems that there is an array bound problem in func `neo-vm\src\neo-vm\ExecutionEngine.ExecuteInstruction`, case OpCode.PICKITEM.

Could anyone have a look at this problem?

**Comment by shargon** [comment]
Try with
states =new State[1];
states[0] =s ;

**Comment by Tommo-L** [comment]
It works well.
![image](https://user-images.githubusercontent.com/43396869/70698682-ed80c200-1d02-11ea-8998-f05effd8d91d.png)

But the following case will fail.
![image](https://user-images.githubusercontent.com/43396869/70698869-36d11180-1d03-11ea-92d7-6a688ec2b950.png)

Do we need to design a Neo contract language? 😂

**Comment by shargon** [comment]
> Do we need to design a Neo contract language?

Of course not, but for fix it we need to find the root cause

**Comment by Qiao-Jin** [comment]
After refining code to avoid above situation still exception will be thrown in line `BasicMethods.DynamicCallContract dyncall = (BasicMethods.DynamicCallContract)result.ToDelegate();`

```
if (operation == "openChannelMockSet")
{
    BigInteger _disputeTimeout = 1000;
    byte _tokenType = 0;
    byte[][] _peerAddrs = new byte[][] { _peerAddr1, _peerAddr2 };
    BigInteger a = 10000, b = 10000;
    BigInteger[] _deposits = new BigInteger[] { a, b };
    byte[] result = getCelerLedgerMockHash();
    BasicMethods.DynamicCallContract dyncall = (BasicMethods.DynamicCallContract)result.ToDelegate();
    return (bool)dyncall("openChannelMockSet", new object[] { _channelId, _disputeTimeout, _tokenAddress, _tokenType, _peerAddrs, _deposits });
}
```
where _peerAddr1 & _peerAddr2 are pre-defined byte arrays, BasicMethods.DynamicCallContract is defiend as follows:
`public delegate object DynamicCallContract(string method, object[] args);`

**Comment by Qiao-Jin** [comment]
Even such code will throw an error.
```
if (operation == "openChannelMockSet")
{
    //BigInteger _disputeTimeout = 1000;
    //byte _tokenType = 0;
    //byte[][] _peerAddrs = new byte[][] { _peerAddr1, _peerAddr2 };
    //BigInteger a = 10000, b = 10000;
    //BigInteger[] _deposits = new BigInteger[] { a, b };
    byte[] result = getCelerLedgerMockHash();
    BasicMethods.DynamicCallContract dyncall = (BasicMethods.DynamicCallContract)result.ToDelegate();
    //return (bool)dyncall("openChannelMockSet", new object[] { _channelId, _disputeTimeout, _tokenAddress, _tokenType, _peerAddrs, _deposits });
}
```

**Comment by lightszero** [comment]
1.can't got release error on unittest.
2.for default we talk about 3.0,if had some 2.0 problem, write 2.0 in title

**Comment by shargon** [comment]
Any advance?

**Comment by erikzhang** [comment]
I think it's solved by #482.
